### ISSUE TITLE
[Bug] HttpClient in Java SDK example causes Linux servers to become unresponsive (Documentation)

### ISSUE THREAD

**Issue by rowolff** [growthbook/growthbook#986]
### Summary

The new Java SDK and it's implementation example in the [docs](https://docs.growthbook.io/lib/java#using-the-gbcontext-builder) and in [GitHub](https://github.com/growthbook/examples/blob/main/jvm-spring-web/src/main/java/com/example/demo/MainController.java#L34-L37) use HttpClient which can cause Linux servers to become unresponsive when there is a large number of requests, due to "Too many open files".

#### Reason
For each TCP connection, HttpClient creates a file on disk, which (with lots of traffic) can very quickly hit the limit of most Linux distros. Result is that further http requests get rejected by the server:
![image-2023-02-06-17-26-39-396](https://user-images.githubusercontent.com/2241551/217280993-b76f69f9-021e-4cfb-8743-824a62e00fe9.png)

#### Expected Behavior

By changing the example to use e.g. org.springframework.web.reactive.function.client.WebClient the issue no longer occurs, even with a large number of requests (assumption: TCP connections are being reused better)
```Kotlin
val featuresEndpoint = new URI("https://cdn.growthbook.io/api/features/<environment_key>");
val webClient = WebClient.builder().build()
val growthBookApiResponse = webClient
                                .get().uri(featuresEndpoint)
                                .retrieve()
                                .bodyToMono(GrowthBookApiResponse::class.java)
                                .onErrorResume { Mono.empty() }
                                .block()

val featuresJson = JSONObject(growthBookApiResponse?.features).toString()
```

#### Current Behavior

Using the HttpClient example code causes servers to stop responding to http requests under heavy load.

**Comment by tinahollygb** [comment]
Thanks for this report!

The first link looks like an implementation of `java.net.http.HttpClient`. `java.net.http.HttpClient#send()` makes a blocking call. In the case of Spring, your example is likely the better approach to take given that you do not experience the performance issues that using the blocking `java.net.http.HttpClient` in the standard library would cause. The documentation example is supposed to serve as a web framework-agnostic example using classes available in the standard library, but it seems that there are definitely better options depending on the framework. Your experience seems to be consistent with issues others have described when using HttpClient with Spring (https://github.com/spring-projects/spring-framework/issues/22333).

With regards to [your second link you shared](https://github.com/growthbook/examples/blob/main/jvm-spring-web/src/main/java/com/example/demo/MainController.java#L34-L37) – it appears to be referencing the head of main which is pointing to an `@Autowired` class which was merged in yesterday. I'm assuming it's the wrong reference and that [this is the reference you meant to link](https://github.com/growthbook/examples/blob/0a823f68154819a06fa617d489c6093a664217b8/jvm-spring-web/src/main/java/com/example/demo/MainController.java#L34-L37) (rather than those same lines at the head of main). 

Yesterday, we released 0.4.0 which contains a new class `GBFeaturesRepository` which abstracts networking. Under the hood it is using Square's OkHttp as a networking library. We are still working on the official documentation, but if you're interested in it, for now you can check the following links to get started while we prepare the official documentation:

- [JavaDoc for GBFeaturesRepository](https://growthbook.github.io/growthbook-sdk-java/growthbook/sdk/java/GBFeaturesRepository.html)
- Implementations of `@Autowired @Service` classes:
    - [Unencrypted features](https://github.com/growthbook/examples/blob/d9ba3c9a0142cdeffa920ca93115f240166a9e79/jvm-spring-web/src/main/java/com/example/demo/services/AcmeDonutsFeatureService.java) 
    - [Encrypted features](https://github.com/growthbook/examples/blob/d9ba3c9a0142cdeffa920ca93115f240166a9e79/jvm-spring-web/src/main/java/com/example/demo/services/BasicEncryptedFeaturesService.java)
    - [Getting the features and providing them to the GBContext](https://github.com/growthbook/examples/blob/d9ba3c9a0142cdeffa920ca93115f240166a9e79/jvm-spring-web/src/main/java/com/example/demo/MainController.java#L301)
    - Initializing with the builder ([encrypted features](https://github.com/growthbook/growthbook-sdk-java/blob/main/lib/src/test/java/growthbook/sdk/java/GBFeaturesRepositoryTest.java#L54-L58), [unencrypted features](https://github.com/growthbook/growthbook-sdk-java/blob/main/lib/src/test/java/growthbook/sdk/java/GBFeaturesRepositoryTest.java#L43-L46))

If you do try out the new `GBFeaturesRepository`, please let us know how this impacts the performance. This new class also includes a decrypting implementation and the automated refreshing of the features so your team can make updates in the dashboard and see those changes reflected in the app within a short period of time.

We are working on updating the official documentation for the `GBFeaturesRepository` class but for now you may find those links helpful. This feature is available in 0.4.0.

**Comment by tinahollygb** [comment]
We've added some documentation for using the above-mentioned class: https://docs.growthbook.io/lib/java#fetching-cacheing-and-refreshing-features-with-gbfeaturesrepository

We've also added a note regarding choosing the HTTP client for your given framework.

<img width="864" alt="image" src="https://user-images.githubusercontent.com/113377031/218883828-465cdc52-b5f4-48cd-9685-7c51207387ec.png">

Closing this issue. Feel free to open a new issue if you experience the same performance issues with the provided `GBFeaturesRepository` class.
